Added An Overview of Chalk

sunjay opened the pull request:
This is the overview of [chalk](https://github.com/rust-lang-nursery/chalk) given to me by Niko in the #wg-traits Discord channel.

r? @nikomatsakis 

[Rendered](https://github.com/sunjay/rustc-guide/blob/8cfd5c5aede5fe17cc286c230ccd2eba4cc966bd/src/chalk-overview.md)

Even more rendered: (click for full size)

![screen shot 2018-05-14 at 14 44 47-fullpage](https://user-images.githubusercontent.com/530939/40025112-75e93ae0-5785-11e8-86e0-1297dda63565.png)

mark-i-m commented on src/chalk-overview.md:
```diff
@@ -0,0 +1,141 @@
+> Chalk is under heavy development, so if any of these links are broken or if
+> any of the information is inconsistent with the code or outdated, please open
+> an issue so we can fix it. If you are able to fix the issue yourself, we would
+> love your contribution!
```
Maybe a link to the repo here?

mark-i-m commented on src/chalk-overview.md:
```diff
@@ -0,0 +1,141 @@
+> love your contribution!
+
+[Chalk][chalk] recasts Rust's trait system explicitly in terms of logic
+programming by "lowering" Rust code into a kind of logic program we can then
```
Perhaps a link to the lowering chapters would be good here?
- https://rust-lang-nursery.github.io/rustc-guide/traits-lowering-to-logic.html
- https://rust-lang-nursery.github.io/rustc-guide/traits-lowering-rules.html

mark-i-m commented on src/chalk-overview.md:
```diff
@@ -0,0 +1,141 @@
+[Chalk][chalk] recasts Rust's trait system explicitly in terms of logic
+programming by "lowering" Rust code into a kind of logic program we can then
+execute queries against. Its goal is to be an executable, highly readable
+specification of the Rust trait system.[^negativechalk]
```
I don't understand the footnote here?

sunjay replied:
This description is a paraphrased (and copied) version of the blog post in the footnote so I thought I should cite my sources :stuck_out_tongue_closed_eyes:

mark-i-m commented on src/chalk-overview.md:
```diff
@@ -0,0 +1,141 @@
+* [Cyclic queries in chalk](http://smallcultfollowing.com/babysteps/blog/2017/09/12/tabling-handling-cyclic-queries-in-chalk/)
+* [An on-demand SLG solver for chalk](http://smallcultfollowing.com/babysteps/blog/2018/01/31/an-on-demand-slg-solver-for-chalk/)
+
+## Parsing
```
Perhaps add a note that this is completely orthogonal to the Rust AST and grammar.

mark-i-m commented on src/chalk-overview.md:
```diff
@@ -0,0 +1,141 @@
+The parser takes that syntax and produces an [Abstract Syntax Tree (AST)][ast].
+You can find the [complete definition of the AST][chalk-ast] in the source code.
+
+The syntax contains things from Rust that we know and love for example traits,
```
Comma needed after "love"

mark-i-m commented on src/chalk-overview.md:
```diff
@@ -0,0 +1,141 @@
+## Lowering
+
+After parsing, there is a "lowering" phase. This aims to convert traits/impls
+into "program clauses". A [`ProgramClause` (source code)][programclause] is
```
Perhaps a link to https://rust-lang-nursery.github.io/rustc-guide/traits-goals-and-clauses.html would also be useful here?

sunjay replied:
This is done. There is a "See also" link after the bullets below.

mark-i-m commented on src/chalk-overview.md:
```diff
@@ -0,0 +1,141 @@
+  "if" and `conditions = cond1 && cond2 && ...`
+* A universally quantified clause of the form
+  `forall<T> { consequence :- conditions }`
+  * `forall<T> { ... }` is used to represent [universal quantification]. See the
```
Perhaps "universal quantification" would be a good term to add to the glossary?

sunjay replied:
Maybe? It isn't exactly Rust specific, but we certainly could add it if you think it would be useful. :)

mark-i-m commented on src/chalk-overview.md:
```diff
@@ -0,0 +1,141 @@
+    over traits, only types and regions (lifetimes). That is, you can't make a
+    rule like `forall<Trait> { u32: Trait }` which would say "`u32` implements
+    all traits". You can however say `forall<T> { T: Trait }` meaning "`Trait`
+    is implemented by all types".
```
:+1: This is a great example

sunjay replied:
Thanks! All of this is from @nikomatsakis, I just typed it up and made it look pretty :smile:

mark-i-m commented on src/chalk-overview.md:
```diff
@@ -0,0 +1,141 @@
+  * `forall<T> { ... }` is represented in the code using the [`Binders<T>`
+    struct][binders-struct].
+
+This is the phase where we encode the rules of the trait system into logic. For
```
s/This/Lowering/

mark-i-m commented on src/chalk-overview.md:
```diff
@@ -0,0 +1,141 @@
+    struct][binders-struct].
+
+This is the phase where we encode the rules of the trait system into logic. For
+example, if we have:
```
"if we have" -> "if we have the following Rust"

mark-i-m commented on src/chalk-overview.md:
````diff
@@ -0,0 +1,141 @@
+impl<T: Clone> Clone for Vec<T> {}
+```
+
+We generate:
````
"We generate" -> "We generate the following program clause"

mark-i-m commented on src/chalk-overview.md:
```diff
@@ -0,0 +1,141 @@
+formedness" checks. See the [source code][well-formedness-checks] for where
+those are done. The call to `record_specialization_priorities` checks
+"coherence" which means that it ensures that two impls of the same trait for the
+same type cannot exist.
```
I'm guessing that this actually checks that specialization rules are not violated (i.e. overlapping impls are "valid")? In other words, if we wanted to implement specialization, would we just change the implementation of `record_specialization_priorities`?

sunjay replied:
@nikomatsakis can answer this for sure, but my understanding is that this function is named this way because it will eventually help with specialization even though that isn't fully implemented yet. The description about the coherence checks is still correct.

mark-i-m commented on src/chalk-overview.md:
```diff
@@ -0,0 +1,141 @@
+string in the AST but in `ir::Program`, we use numeric indices (`ItemId`).
+
+In addition to `ir::Program` which has "rust-like things", there is also
+`ir::ProgramEnvironment` which is "pure logic". The main field in that is
```
"that" -> "that struct"?

mark-i-m commented on src/chalk-overview.md:
```diff
@@ -0,0 +1,141 @@
+## Rules
+
+The `rules` module works by iterating over every trait, impl, etc. and emitting
+the rules that come from each one. The traits section of the rustc-guide (that
```
Perhaps link to https://rust-lang-nursery.github.io/rustc-guide/traits-lowering-rules.html for convenience?

mark-i-m commented on src/chalk-overview.md:
```diff
@@ -0,0 +1,141 @@
+## Testing
+
+TODO: Basically, [there is a macro](https://github.com/rust-lang-nursery/chalk/blob/94a1941a021842a5fcb35cd043145c8faae59f08/src/solve/test.rs#L112-L148)
+that will take syntax and run it through the full pipeline described above.
```
Rust or chalk syntax? (I'm guessing the latter)

sunjay commented:
@mark-i-m I believe I have addressed all of your comments. GitHub's UI is weird and it hid some of them from me so it took a few tries to get all of them. Let me know if there are any more that I missed or if you have any more feedback. Thanks for your review!

mark-i-m commented:
LGTM :) Thanks!

I would be fine with merging this now, but if you want, we can also wait for a review from Niko.

sunjay commented:
Sure. Let's wait for Niko to see if everything is accurately written. :)

nikomatsakis commented:
Looks good to me!
